javalib/src/main/scala/java/lang/dec2flt/Bellerophon.scala

```diff
@@ -0,0 +1,288 @@
+  private final val ExtendedMaxSig = BigInteger.ZERO.setBit(ExtendedSigBits)
+
+  def bellerophonFloat(f: BigInteger, e: Int): scala.Float =
+    bellerophon(f, e, Binary32)
+
+  def bellerophonDouble(f: BigInteger, e: Int): scala.Double =
+    bellerophon(f, e, Binary64)
+
+ /** An implementation of Bellerophon algorithm from
+  *  "How to Read Floating Point Numbers Accurately" by William D. Clinger.
+  *
+  *  The Bellerophon algorithm efficiently finds the best binary
+  *  approximation for a decimal number `f * 10^e`.
+  *  While an iterative approach like AlgorithmM (defined in the paper)
+  *  can find the best approximation by repeatedly using slow
+  *  BigInteger arithmetic, Bellerophon provides non-iterative approach.
+  *
+  *  The basic idea is, instead of directly computing the best approximation
+  *  for `f * 10^e` it uses higher-precision (p-bit (64-bit) significand)
+  *  floating-point numbers to approximate `f` and `10^e` separately.
+  *  Then multiplies these approximations and analyzes the
+  *  error when multiplying and rounding the result to the target
+  *  n-bit (n-bit significand) precision.
+  *
+  *  In some cases, the calculated approximation isn't the best, and in that case
+  *  the algorithm falls back to AlgorithmR.
+  *  This algorithm takes a good appoximation as a starting point,
+  *  and iteratively finds the best approximation.
+  *  Since the calculated approximation by Bellerophon is very close to the best
+  *  it converges in just a few steps.
+  *
+  *  For extreme values that would overflow to infinity or underflow to a
+  *  subnormal number or zero, neither the Bellerophon nor AlgorithmR does work.
+  *  In these cases, the algorithm falls back to the original, slower AlgorithmM
+  *  with the modification mentioned in section 8 of the paper.
+  *  > With IEEE arithmetic, for example, a denormalized result
+  *  > may be required. Denormalized results can be generate
+  *  > by a modified form of AlgorithmM that terminates immediately
+  *  > when the minimum exponent is reached.
+  */
+  private def bellerophon(f: BigInteger, e: Int, fmt: FloatingPointFormat): fmt.Repr = {
+    // Fast path: when f < 2^sigbits (2^53 for Double), and |e| < ceil(log5(2^sigbits)),
+    // both f and 10^e can be represented perfectly in target precision without error.
+    //
+    // Since 10^e = 5^e * 2^e, the 2^e part can be absorbed into the binary exponent.
+    // Therefore, the 10^e can be represented without error if 5^e fits within the significand.
+    // This condition is expressed as 5^e < 2^sigbits <=> e < log5(2^sigbits).
+    // As `e` is an integer, the condition is equivalent to e < ceil(log5(2^sigbits)).
+    if (f.compareTo(BigInteger.valueOf(fmt.MaxSig)) < 0 && e >= 0 && e < fmt.CeilLog5OfMaxSig) {
+      fmt.mul(fmt.fromLong(f.longValueExact()), fmt.powerOfTen(e))
+    } else if (f.compareTo(BigInteger.valueOf(fmt.MaxSig)) < 0 && e < 0 && -e < fmt.CeilLog5OfMaxSig) {
```

Comment:
`f.compareTo(BigInteger.valueOf(fmt.MaxSig))` is done twice. Do it only once before the `if`s.

Also, consider storing `BigInteger.valueOf(fmt.MaxSig)` as an instance property of `fmt`.